traceapp/tmpl/data/trace.html

```diff
@@ -420,6 +428,94 @@
     if(data != null) {
       timelineHover();
     }
+
+    // save state for last sort
+    var lastSort = {
+      name: '',
+      order: false,
+      toggleOrder: function(name){
+        if (this.name !== name) {
+          this.name = name;
+          this.order = false;
+          return 1;
+        }
+
+        this.order = !this.order;
+        if (this.order) {
+          return -1;
+        }
+        return 1;
+      }
+    };
+
+    // sort by starting_time
+    function toggleSortStartTime() {
+      var sortOrder = lastSort.toggleOrder('start_time');
+      data.sort(function(a, b) {
+        // validation
+        if (typeof a.times !== 'object' || typeof b.times !== 'object') return -1;
+        if (a.times.length === 0 || b.times.length === 0) return -1;
+
+        // sort
+        if (a.times[0].starting_time < b.times[0].starting_time) return sortOrder * -1;
+        if (a.times[0].starting_time > b.times[0].starting_time) return sortOrder;
+        return 0
+      })
+      timelineHover();
+    }
+
+    // sort by ending_time
+    function toggleSortEndTime() {
+      var sortOrder = lastSort.toggleOrder('end_time');
+      data.sort(function(a, b) {
+        // validation
+        if (typeof a.times !== 'object' || typeof b.times !== 'object') return -1;
+        if (a.times.length === 0 || b.times.length === 0) return -1;
+
+        // sort
+        if (a.times[0].ending_time < b.times[0].ending_time) return sortOrder * -1;
+        if (a.times[0].ending_time > b.times[0].ending_time) return sortOrder;
```

Comment:
> Could be written as an else statement, instead of doing the extra comparison, right?

thanks, I changed the style to use `else if`.

> Also, what is the sortOrder * -1 needed for?

`sort()` might returns 3 value.
- `-1` when less than
- `1` when greater than
- `0` when equal value

[from the document of `Array.prototype.sort()`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/sort), example sort code is like below,

```js
function compare(a, b) {
  if (a is less than b by some ordering criterion) {
    return -1;
  }
  if (a is greater than b by the ordering criterion) {
    return 1;
  }
  // a must be equal to b
  return 0;
}
```

here,  `sortOrder` should be `1` or `-1`
-  `1` for ascending order(when `lastSort.descendingOrder === false`)
-  `-1` for descending order(when `lastSort.descendingOrder === true`)

and switching the order inside of sort function to multiply `-1` to use `sortOrder = 1 or -1`

```js
// combine these two function into one to use `sorOrder` variable

// ascending order
function compare(a, b) {
  if (a is less than b by some ordering criterion) {
    return -1;
  }
  if (a is greater than b by the ordering criterion) {
    return 1;
  }
  // a must be equal to b
  return 0;
}

// descending order
function descendingCompare(a, b) {
  if (a is less than b by some ordering criterion) {
    return 1; // inversed from ascending
  }
  if (a is greater than b by the ordering criterion) {
    return -1; // inversed from ascending
  }
  // a must be equal to b
  return 0;
}
```